Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md

```diff
@@ -785,31 +785,55 @@ You can detect popular reverse engineering tools that have been installed in an
-```c
+With API Level 25 and below it was possible to query for all running services by using the Java method  (getRunningServices[https://developer.android.com/reference/android/app/ActivityManager.html#getRunningServices(int) "getRunningServices"]. This allows to iterate through the list of running UI activities, but will not show you daemons like the frida-server. Starting with API Level 26 and above `getRunningServices()` will even only return the caller's own services.
+
+A working solution to detect the frida process is to us the command `ps` instead.
+
+```Java
 public boolean checkRunningProcesses() {
 
-  boolean returnValue = false;
+    boolean returnValue = false;
+
+    try {
+        Process process = Runtime.getRuntime().exec("ps");
+        BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+        int read;
+        char[] buffer = new char[4096];
+        StringBuffer output = new StringBuffer();
+        while ((read = reader.read(buffer)) > 0) {
+            output.append(buffer, 0, read);
+        }
+        reader.close();
 
-  // Get currently running application processes
-  List<RunningServiceInfo> list = manager.getRunningServices(300);
+        // Waits for the command to finish.
+        process.waitFor();
+        Log.d("fridaserver", output.toString());
 
-  if(list != null){
-    String tempName;
-    for(int i=0;i<list.size();++i){
-      tempName = list.get(i).process;
+        if(output.toString().contains("frida-server")) {
+            Log.d("fridaserver","Frida Server process found!" );
+            returnValue = true;
+        }
+
+    } catch (IOException e) {
+
+    } catch (InterruptedException e) {
 
-      if(tempName.contains("fridaserver")) {
-        returnValue = true;
-      }
     }
-  }
-  return returnValue;
+
+    return returnValue;
 }
+
 ```
 
-This works if Frida is run in its default configuration. Perhaps it's also enough to stump some script kiddies during their first steps in reverse engineering. It can, however, be easily bypassed by renaming the frida-server binary, so we should find a better method.
+Starting with Android Nougat (API Level 24) the `ps` command will only return processes started by the user itself. When executing `ps` it will read the information from `/proc` and it's not possible to access information that belongs to other user ids.
+
+![Executing ps on Android Lollipop](Images/Chapters/0x05j/Android_Lollipop_ps.png)
+
+![Executing ps on Android Nougat](Images/Chapters/0x05j/Android_Nougat_ps.png)
+
+Even if the process name could easily be detected, this would only work if Frida is run in its default configuration. Perhaps it's also enough to stump some script kiddies during their first steps in reverse engineering. It can, however, be easily bypassed by renaming the frida-server binary. So because of this and the technical limitations of querying the process names in recent Android versions, we should find a better method.

```

Comment:
Additionally: you know that there will be a gadget in your patched runtime. This means that the app has been resigned with either the same cert and a different signature or with another cert with a different signature. Can you refer to integrity checking, or update the write up about that given v2/v3 signign scheme in android?
Because that way you can see that your binary has been modified.

Reply:
I was already thinking about integrity checking of the APK to verify if the APK was tampered . through repackaging. let me check if this is already described in the reverse engineering part.

Reply:
I made a very simple check. Let me know  what you think